### ISSUE TITLE
cinnamon: Reloading desklets leaves them unresponsive for several seconds

### ISSUE THREAD

**Issue by haggen88** [linuxmint/mint21.3-beta#26]
After using the 'Reload All Desklet' action, they freeze and do not respond for several seconds (at least 30 seconds in my case) or until a window near the desklet is opened or occupied (or cinnamon is restarted).

[desklet reload accion issue.webm](https://github.com/linuxmint/mint21.3-beta/assets/68134588/6598bbc4-09c2-4e0d-9b48-b07ca1a102e3)

This test was in live-usb mode, I did not have a significant cpu or ram overload, I also observed that with 4 desklet cinnamon it froze.

logs: [logs.zip](https://github.com/linuxmint/mint21.3-beta/files/13690930/logs.zip)
sys info: [sys info.zip](https://github.com/linuxmint/mint21.3-beta/files/13690945/sys.info.zip)

**Comment by rcalixte** [comment]
This should probably be moved as an issue to [linuxmint/cinnamon-spices-actions](https://github.com/linuxmint/cinnamon-spices-actions) as it is for the Spice itself. (I don't think this impacts Cinnamon 5.8 or 6.0 differently.)

**Comment by rcalixte** [comment]
I can replicate this in a live environment. Normal operations for mouse events to the desklets resumes after the reload only after focus is switched to another window, either by launching a new window or by just switching to an open window.

**Comment by rcalixte** [comment]
There is nothing of value in the logs and as of right now, the behavior is only isolated to live environments. (Cinnamon also crashed quite a bit on consecutive reloads.)

**Comment by rcalixte** [comment]
@haggen88 If I move the desklets before trying to reload them, I can get a successful reload. Can you replicate this?

**Comment by rcalixte** [comment]
@haggen88 Removing cpuload@kimse solved the issue for me. I'll move to exclude this from desklets that are reloaded (and implement it in a way we can add more if there are more problematic ones identified).

**Comment by rcalixte** [comment]
Thank you for reporting this @haggen88! The latest version should work without issue. I need to find out what specifically made cpuload@kimse problematic and either identify potential other desklets or find some other permanent fix. I checked the most popular desklets by rating and those all seem fine. I'll make a note for myself to look into this further.

**Comment by haggen88** [comment]
Hello @rcalixte, I just tried in live-usb mode and the same thing happens this time I used a single desklet (diskspace@schorschii):[logs.zip](https://github.com/linuxmint/mint21.3-beta/files/13693733/logs.zip)

I don't know if it is a particular desklet (or the live-usb mode), that is the reason for reporting it in this beta.

**Comment by haggen88** [comment]
oops. click error

**Comment by rcalixte** [comment]
That was one of the ones I'd tested. I also have it on all of my machines with Cinnamon. I'll take a look in a little bit. It also looks like it was the only desklet that was added.

**Comment by anaximeno** [comment]
That might be a problem that happens with other desklets as well. Sometimes, after dragging some desklets to other positions, they become unresponsive to dragging, but if you try to drag it again while clicking 'alt' on the keyboard it works again. it doesn't happen always, but it happens sometimes. The problem I related here happened in other versions also, so it's not a specific regression of this new version (for the case I related).

**Comment by anaximeno** [comment]
Anyway, should check for errors in the `~/.xsession-errors` also while this problem is happening to see if something is being logged there.

**Comment by rcalixte** [comment]
There isn't much in the logs. I can reproduce that the desklets are unresponsive until any other window is activated. I had a terminal window open but unfocused and just mousing over it without giving it focus returned the responsiveness to the desklets. This doesn't work with switching on empty desktops though. This leads me to thinking it might be something in the compositor that has always been there but is being exacerbated by the Action. I'm also not able to replicate this in the Wayland session of the live ISO. (Can you test in the Wayland session and confirm that it's fine?) I've had times in the past where I had stuck desklets that I resolved by restarting Cinnamon. This feels like it will be tricky to resolve in the short-term though.

**Comment by anaximeno** [comment]
@rcalixte two things:
1. I also was not able to reproduce the bug under Wayland;
2. The issue won't happen if you directly run the script inside the action you created to reload the desklets on the terminal. However, if instead, you add a sleep command to delay the execution of the script and then minimize the terminal window (so the focus would be on the desktop), the problem would happen again. cmd: `sleep 3 && reload-all-desklets@rcalixte.py`

> This leads me to thinking it might be something in the compositor that has always been there but is being exacerbated by the Action.

It seems to be something like that.

Maybe try to add a loading window to gain focus during the loading time?

**Comment by rcalixte** [comment]
> 2\. The issue won't happen if you directly run the script inside the action you created to reload the desklets on the terminal.

This is likely because of the focus of the terminal window.


> Maybe try to add a loading window to gain focus during the loading time?

This is brilliant. I wish I'd thought of it myself. I'll open a pull request for this after testing in both types of sessions.

**Comment by rcalixte** [comment]
@haggen88 @anaximeno Thanks for the help with this! Please test the latest version and confirm it's working for you!

**Comment by rcalixte** [comment]
I think I let my excitement get the best of me here... This may need another look. The dialog doesn't seem to be consistently displaying for me.

**Comment by anaximeno** [comment]
> I think I let my excitement get the best of me here... This may need another look. The dialog doesn't seem to be consistently displaying for me.

Yes, I had the same problem while testing, I believe the reload is way faster than the time it takes to open the zenity dialog, I thought of a solution for that, instead of showing a loading window, show an information dialog, in the end, informating the user that the reload has finished.

**Comment by anaximeno** [comment]
Also, you can remove the cpuload desklet from the BADS as it's not the cause of the problem as we noticed.

**Comment by rcalixte** [comment]
> Also, you can remove the cpuload desklet from the BADS as it's not the cause of the problem as we noticed.

In my tests, that desklet still causes Cinnamon to crash when it's attempted to be reloaded.


> Yes, I had the same problem while testing, I believe the reload is way faster than the time it takes to open the zenity dialog, I thought of a solution for that, instead of showing a loading window, show an information dialog, in the end, informating the user that the reload has finished.

I think the subprocess call is actually having issues with the zenity process. I'm taking a look at why.

**Comment by anaximeno** [comment]
> In my tests, that desklet still causes Cinnamon to crash when it's attempted to be reloaded.

Oh, ok.

> I think the subprocess call is actually having issues with the zenity process. I'm taking a look at why.

I did some tests and the problem was not with zenity (or subprocess), the `os.kill(process.pid, signal.SIGTERM)` was being called before zenity could even open the window. So instead, I tried to remove the `with` and the `os.kill`, and just added a `subprocess.run` to create an info window in the end (after the for loop) informing the user that the reload finished and it worked (as the goal is just to shift the focus from the desktop).

**Comment by rcalixte** [comment]
> I did some tests and the problem was not with zenity (or subprocess), the `os.kill(process.pid, signal.SIGTERM)` was being called before zenity could even open the window. So instead, I tried to remove the `with` and the `os.kill`, and just added a `subprocess.run` to create an info window in the end (after the for loop) informing the user that the reload finished and it worked (as the goal is just to shift the focus from the desktop).

That's because of the syntax to the Popen. It's a syntax issue that needs to be debugged.

**Comment by anaximeno** [comment]
That was what I suspected at first, so I also ran that code on the Python interpreter to check that before and made some tests involving sleep to confirm that the problem was the timing, etc. But, anyway, if you find the problem just let me know.

**Comment by rcalixte** [comment]
This should be safe to close now. Quite the journey. 😅

(There is technically an exception generated once the dialog is closed but this is a much lower priority.)

**Comment by anaximeno** [comment]
Tested under Wayland and X11, and it's working!

**Comment by rcalixte** [comment]
I just pushed two more minor changes:

* The dialog will now only show in X11 sessions if desklets are not locked
* Fixed the GTK exception that was being generated

**Comment by haggen88** [comment]
I confirm, no problems in x11

@rcalixte  The only thing is that the message in the window still appears in English despite having a Spanish translation. I guess it's a live-usb thing. 
Thanks!

**Comment by rcalixte** [comment]
> @rcalixte The only thing is that the message in the window still appears in English despite having a Spanish translation. I guess it's a live-usb thing. Thanks!

No, it's another bug from me. 😔 
Thanks for catching it! I'm pushing out a fix now!
